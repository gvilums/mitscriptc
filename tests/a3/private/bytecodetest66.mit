calls = 0;
hi = fun(x, y) {
    global calls;
    out = x + calls;
    print(out);
    calls = calls + 1;
    return(out);
    print("lol");
    return(1/0);
};

hi(hi(hi("5", 1), hi("6", 1)), "3");

//rec = {dog: hi("a", 1); cat: hi("b", 2); dog: hi("c", 3);};
//print(rec);

oldprint = print;
print = fun(x) {
    oldprint("darn");
    oldprint(x);
};
print("test");
print("test");
print("test");
hi("foo", 0);

i = 0;
while(i < 999) {
    i = i + 1;
}
z = 0;
foo = fun() {
    while(z < 10) {
        return 0;
        print("oops");
    }
};
foo();
foorec = {x: 0; a: 3;};
print(foorec["x"]);
print("ok");
