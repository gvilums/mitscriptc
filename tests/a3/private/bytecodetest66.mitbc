function
{
	functions =
	[
		function
		{
			functions = [],
			constants = [],
			parameter_count = 1,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [],
			parameter_count = 0,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [],
			parameter_count = 1,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [],
			instructions = 
			[
			]
		},
		function
		{
			functions = [],
			constants = [None, 1, "lol", 0],
			parameter_count = 2,
			local_vars = [x, y, out],
			local_ref_vars = [],
			free_vars = [],
			names = [calls, print],
			instructions = 
			[
				load_const	0
				store_local	2
				load_local	0
				load_global	0
				add
				store_local	2
				load_global	1
				load_local	2
				call	1
				pop
				load_global	0
				load_const	1
				add
				store_global	0
				load_local	2
				return
				load_global	1
				load_const	2
				call	1
				pop
				load_const	1
				load_const	3
				div
				return
			]
		},
		function
		{
			functions = [],
			constants = ["darn", None],
			parameter_count = 1,
			local_vars = [x],
			local_ref_vars = [],
			free_vars = [],
			names = [oldprint],
			instructions = 
			[
				load_global	0
				load_const	0
				call	1
				pop
				load_global	0
				load_local	0
				call	1
				pop
				load_const	1
				return
			]
		},
		function
		{
			functions = [],
			constants = [10, 0, "oops", None],
			parameter_count = 0,
			local_vars = [],
			local_ref_vars = [],
			free_vars = [],
			names = [z, print],
			instructions = 
			[
				load_global	0
				load_const	0
				swap
				gt
				not
				if	8
				load_const	1
				return
				load_global	1
				load_const	2
				call	1
				pop
				goto	-12
				load_const	3
				return
			]
		}
	],
	constants = [0, "5", 1, "6", "3", "test", "foo", 999, 3, "x", "ok"],
	parameter_count = 0,
	local_vars = [],
	local_ref_vars = [],
	free_vars = [],
	names = [calls, hi, oldprint, print, i, z, foo, foorec, x, a, input, intcast],
	instructions = 
	[
		load_func	0
		alloc_closure	0
		store_global	3
		load_func	1
		alloc_closure	0
		store_global	10
		load_func	2
		alloc_closure	0
		store_global	11
		load_const	0
		store_global	0
		load_func	3
		alloc_closure	0
		store_global	1
		load_global	1
		load_global	1
		load_global	1
		load_const	1
		load_const	2
		call	2
		load_global	1
		load_const	3
		load_const	2
		call	2
		call	2
		load_const	4
		call	2
		pop
		load_global	3
		store_global	2
		load_func	4
		alloc_closure	0
		store_global	3
		load_global	3
		load_const	5
		call	1
		pop
		load_global	3
		load_const	5
		call	1
		pop
		load_global	3
		load_const	5
		call	1
		pop
		load_global	1
		load_const	6
		load_const	0
		call	2
		pop
		load_const	0
		store_global	4
		load_global	4
		load_const	7
		swap
		gt
		not
		if	6
		load_global	4
		load_const	2
		add
		store_global	4
		goto	-10
		load_const	0
		store_global	5
		load_func	5
		alloc_closure	0
		store_global	6
		load_global	6
		call	0
		pop
		alloc_record
		dup
		dup
		load_const	0
		field_store	8
		load_const	8
		field_store	9
		store_global	7
		load_global	3
		load_global	7
		load_const	9
		index_load
		call	1
		pop
		load_global	3
		load_const	10
		call	1
		pop
	]
}