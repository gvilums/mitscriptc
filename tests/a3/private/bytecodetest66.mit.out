50
61
502
5023
darn
test
darn
test
darn
test
darn
foo4
darn
0
darn
ok
